File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        substring_set = set()
        max_len = 0

        for r in range(len(s)):
            while s[r] in substring_set:
                substring_set.remove(s[l])
                l += 1
            
            substring_set.add(s[r])
            max_len = max(max_len,r-l+1)
        return max_len
        

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int L=0;
        int max = 0;

        for(int R=0;R<s.length();R++){
            while(set.contains(s.charAt(R))){
                set.remove(s.charAt(L));
                L++;
            }
            set.add(s.charAt(R));
            max = Math.max(max,set.size());
        }
        return max;
    }
}